### PR TITLE
BLD: use macos-11 image on azure, macos-1015 is deprecated

### PR DESCRIPTION
**Pull request by mattip**
There is [a warning](https://dev.azure.com/numpy/numpy/_build/results?buildId=25324&view=logs&j=9d7fb089-bdae-57e1-a4ec-9025f62c646c) in the CI runs that the macos-1015 image is deprecated.

### CODE REVIEW

**Comment by rgommers** [comment]
Good idea to update. Some of the `f2py` tests do not seem happy. 

The `pytest` output is pretty terrible, so it's not clear from the summary which 6 tests are failing - I only see a long list of `XFAIL` and `SKIPPED`.

**Comment by mattip** [comment]
The [error](https://dev.azure.com/numpy/numpy/_build/results?buildId=25367&view=logs&j=5a9ef69c-845d-5ef3-f457-68d7d7acc0d6&t=b19c91e2-8e9c-5ce5-8245-4675462d2d3c&l=80295) is a failure to compile. Do the tests work locally on macOS 11?
> `ld: library not found for -lSystem`

**Comment by rgommers** [comment]
> The [error](https://dev.azure.com/numpy/numpy/_build/results?buildId=25367&view=logs&j=5a9ef69c-845d-5ef3-f457-68d7d7acc0d6&t=b19c91e2-8e9c-5ce5-8245-4675462d2d3c&l=80295) is a failure to compile. Do the tests work locally on macOS 11?

Yes, but macOS 11 isn't really a thing many people use; it was followed by 12.0 shortly after, and we're at 12.4 now.

There is gh-20677 which looks like the same problem. However, it's kind of weird in this PR that the regular build works but the test suite fails on compiling something. So perhaps it has to do with the way the `f2py` tests are run, since a linker flag goes missing.

**Comment by mattip** [comment]
What would be the downside of moving to a `macos-12`-based image? Does that mean the wheels cannot be used on macos < 12?

**Comment by rgommers** [comment]
For `arm64` wheels it is fine to require macOS >= 12.0, because 11.0 was quite buggy and short-lived. A number of fixes that we needed in numpy for scipy ended up in macOS 12.1 and the relevant Clang versions. SciPy already requires >= 12.0; I can't imagine that there are many `arm64` users on 11 still.

**Comment by mattip** [comment]
Even with macOS-12 the [error](https://dev.azure.com/numpy/numpy/_build/results?buildId=25547&view=logs&j=9d7fb089-bdae-57e1-a4ec-9025f62c646c&t=a110bc30-f03d-50a4-8dba-c2547e2417d7&l=557) of `ld: library not found for -lSystem` is still there. Here is the compiler command:
```
INFO: Fortran f77 compiler: /usr/local/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -fPIC -O3 -funroll-loops
E                                  Fortran f90 compiler: /usr/local/bin/gfortran -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
E                                  Fortran fix compiler: /usr/local/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
E                                  INFO: compile options: '-I/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/src.macosx-10.15-x86_64-3.8 -I/Users/runner/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/numpy-1.24.0.dev0+638.g596fb0dcf-py3.8-macosx-10.15-x86_64.egg/numpy/core/include -I/Users/runner/hostedtoolcache/Python/3.8.13/x64/include/python3.8 -c'
E                                  extra options: '-J/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/ -I/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/'
E                                  INFO: gfortran:f90: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/src.macosx-10.15-x86_64-3.8/_test_abstract_interface_TestAbstractInterface_ext_module-f2pywrappers2.f90
E                                  INFO: /usr/local/bin/gfortran -Wall -g -Wall -g -undefined dynamic_lookup -bundle /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/src.macosx-10.15-x86_64-3.8/_test_abstract_interface_TestAbstractInterface_ext_modulemodule.o /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/src.macosx-10.15-x86_64-3.8/fortranobject.o /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmpg01ush4w/foo.o /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmp3ofky4jb/src.macosx-10.15-x86_64-3.8/_test_abstract_interface_TestAbstractInterface_ext_module-f2pywrappers2.o -L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin13/4.9.0 -L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin13/4.9.0/../../.. -L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin13/4.9.0/../../.. -L/Users/runner/hostedtoolcache/Python/3.8.13/x64/lib -lgfortran -o ./_test_abstract_interface_TestAbstractInterface_ext_module.cpython-38-darwin.so
E                                  ld: library not found for -lSystem

```

**Comment by rgommers** [comment]
I think you can go back to `macos-11` then. We are doing something fishy somewhere to make `-lSystem` appear; I can find very few other references for this error. 

This looks suspicious but is no longer used, so doesn't seem like the problem:
```
[ -n "$USE_XCODE_10" ] && /bin/bash -c "sudo xcode-select -s /Applications/Xcode_10.app/Contents/Developer"
```
Somehow `-lSystem is only injected for `gfortran` invocations. The code in `gnu.py` appends flags from `sysconfig.get_config_vars()`, how about adding some debug prints to see if in CI any of the entries returned from there include `System`?

**Comment by mattip** [comment]
On a mac, I see `-lSystem` in `CONFIG_ARGS` (not relevant), `LIBS` and `SHLIBS`. But I don't see `LIBS` nor `SHLIBS` used in NumPy. I did find this [stack overflow about `-lSystem` and gfortran](https://stackoverflow.com/questions/56156520/gfortran-error-ld-library-not-found-for-lsystem-when-trying-to-compile), and [another](https://apple.stackexchange.com/questions/408999/gfortran-compiler-error-on-mac-os-big-sur) that suggest some interplay between gcc, gfortran, and xcode.

<details><summary>sysconfig variables</summary>

```
% python3 -m sysconfig |grep "lSystem" 
	CONFIG_ARGS = "'-C' '--host=x86_64-apple-darwin' '--build=x86_64-apple-darwin' '--enable-framework=/Applications/Xcode.app/Contents/Developer/Library/Frameworks' '--with-framework-name=Python3' '--with-openssl=/AppleInternal/BuildRoot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.Internal.sdk/usr/local/libressl-2.8' '--with-openssl-ldflags=-L/AppleInternal/BuildRoot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.Internal.sdk/usr/local/libressl-2.8/lib -F/AppleInternal/BuildRoot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.Internal.sdk/System/Library/PrivateFrameworks' '--with-openssl-libs=-framework TrustEvaluationAgent' '--with-bundleid-prefix=com.apple' '--enable-ipv6' '--prefix=/Applications/Xcode.app/Contents/Developer/usr' '--with-pymalloc' '--with-cxx-main=xcrun -sdk macosx clang -arch arm64 -arch x86_64' 'PYTHON_FOR_BUILD=PYTHONPATH=/AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/python3/python3-110/Python/Lib DYLD_FRAMEWORK_PATH=/AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Binaries/python3/install/TempContent/Objects/host /AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Binaries/python3/install/TempContent/Objects/host/python.exe' 'TRAIN_STYLE=DT' 'CC=xcrun -sdk macosx clang' 'CXX=xcrun -sdk macosx clang++' 'CPP=xcrun -sdk macosx clang -E' 'CFLAGS=-iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration' 'CPPFLAGS=-iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/Headers -arch arm64' 'LIBS=-lSystem' 'LDSHARED=xcrun -sdk macosx clang -bundle -undefined dynamic_lookup -arch arm64 -arch x86_64' 'LDFRAMEWORK=xcrun -sdk macosx clang -dynamiclib -arch arm64 -arch x86_64' 'LDFLAGS=-Wl,-headerpad,0x1000' 'OBJROOT=/AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Binaries/python3/install/TempContent/Objects' 'SDKROOT=/AppleInternal/BuildRoot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.Internal.sdk' 'build_alias=x86_64-apple-darwin' 'host_alias=x86_64-apple-darwin'"
	LIBS = "-ldl -lSystem  -framework CoreFoundation"
	SHLIBS = "-ldl -lSystem  -framework CoreFoundation"
```

</details>

**Comment by rgommers** [comment]
Ah, it depends on the Python install:
```
# conda-forge
% python3 -m sysconfig |grep "SHLIBS"
	SHLIBS = "-ldl   -framework CoreFoundation"

# Apple's Python
% /usr/bin/python3 -m sysconfig |grep "SHLIBS"
	SHLIBS = "-ldl -lSystem  -framework CoreFoundation"
```

**Comment by mattip** [comment]
Trying to reproduce this on a local M1 machine leads me down a rabbit hole. When I build NumPy with `arch -x86_64 python runtests.py`, the dragon4.c code does not compile. Somehow `npy_float128` is not defined.

**Comment by mattip** [comment]
macOS passed in 07a4352. I am not sure why. Trying again with a little less debugging.

**Comment by mattip** [comment]
CI is passing, twice (both 07a4352 and the latest commit)

File: azure-pipelines.yml [hunk 1]
```diff
@@ -205,9 +198,13 @@ stages:
...
       displayName: 'Run Refguide Check'
       condition: eq(variables['USE_OPENBLAS'], '1')
-    - script: python runtests.py -n --mode=full -- -rsx --junitxml=junit/test-results.xml
+    - script: |
+        echo LIBRARY_PATH ${LIBRARY_PATH}
+        python runtests.py -n --mode=full -- -rsx --junitxml=junit/test-results.xml
       displayName: 'Run Full NumPy Test Suite'
       condition: eq(variables['USE_OPENBLAS'], '1')
+      env:
+        LIBRARY_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib
```
Comment (rgommers) [inline azure-pipelines.yml:207]: How about above this line adding:
```bash
# The Python install in GitHub Actions adds `-lSystem`, which fails without adding this (see gh-22043)
```

**Review by rgommers — APPROVED** [review]
LGTM, recommend adding a comment but overall good to merge.

File: azure-pipelines.yml (hunk 1 above)
Comment (mattip) [reply azure-pipelines.yml:208]: I don't think it is coming from the macOS-provided python, rather from gcc/gfortran (as opposed to clang, which does not add it). I did not find SHLIBS in NumPy nor in distutils/setuptools, and the stack overflows I linked to did not mention python. Adding a comment in that vein.

**Review by rgommers — APPROVED** [review]
All green, so in it goes. Thanks @mattip!

File: azure-pipelines.yml (hunk 1 above)
Comment (rgommers) [reply azure-pipelines.yml:208]: makes sense. that's a gfortran that anyway needs an update soon, so hopefully that goes away in the next update

**Comment by rgommers** [comment]
I think this explains what is going on: https://developer.apple.com/forums/thread/655588?answerId=665804022#665804022. The stub library `libSystem.tbd` indeed exists where I'd expect it based on that explanation. 

> I don't think it is coming from the macOS-provided python, rather from gcc/gfortran (as opposed to clang, which does not add it).

From the link above: _OTOH, if your third-party tools are using their own linker then you’ll have to determine whether that linker has been updated to use stub libraries._ So yes, that makes sense. The `gfortran` we use in this job should be removed soon, it's way too old and now also well below the minimum GCC version we need.

**Comment by charris** [comment]
Marked for backport, looks like this will be needed for 1.23.x soon. It would be good if 12.0 could be made to work, I suspect we will need it soon.

**Pull request merged**